src/components/vm/nics/nicBody.tsx

```diff
@@ -271,3 +304,364 @@ export const NetworkTypeAndSourceRow = ({
         </>
     );
 };
+
+export function portForwardText(pf: VMInterfacePortForward): string {
+    let text = "";
+
+    if (pf.address)
+        text += pf.address + ":";
+    if (pf.dev)
+        text += pf.dev + ":";
+    if (!pf.range.some(r => r.exclude != "yes"))
+        text += _("all");
+    for (let i = 0; i < pf.range.length; i++) {
+        const r = pf.range[i];
+        if (i > 0)
+            text += ",";
+        if (r.exclude == "yes")
+            text += _(" excluding ");
+        text += r.start;
+        if (r.end)
+            text += "-" + r.end;
+        if (!r.exclude)
+            text += " → " + (r.to || r.start) + (r.end ? "-" + (Number(r.to || r.start) + Number(r.end) - Number(r.start)).toString() : "");
+    }
+    text += "/" + pf.proto;
+
+    return text;
+}
+
+export function interfacePortForwardsToDialog(portForwards: VMInterfacePortForward[]): DialogPortForward[] {
+    return portForwards.map(pf => {
+        if (!pf.dev && pf.range.length == 1 && pf.range[0].exclude != "yes") {
+            const r = pf.range[0];
+            return {
+                kind: "simple",
+                address: pf.address || "",
+                proto: pf.proto || "",
+                host: r.start + (r.end ? "-" + r.end : ""),
+                guest: r.to || "",
+            };
+        } else {
+            return {
+                kind: "complex",
+                config: pf,
+            };
+        }
+    });
+}
+
+export function dialogPortForwardsToInterface(portForwards: DialogPortForward[]): VMInterfacePortForward[] {
+    // NOTE: We have to set every field here, none of them can be
+    // "null". Our return value is passed to virt-xml, and we want
+    // virt-xml to set everything and not leave anything behind from
+    // the previous portForward XML element.
+    return portForwards.map((pf: DialogPortForward) => {
+        if (pf.kind == "simple") {
+            const range = pf.host.split("-");
+            return {
+                address: pf.address,
+                dev: "",
+                proto: pf.proto,
+                range: [
+                    {
+                        start: range[0],
+                        end: range[1] || "",
+                        to: pf.guest,
+                        exclude: "",
+                    }
+                ],
+            };
+        } else {
+            return pf.config;
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2331-790d4f2a-20251111-065830-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)